File: textbook/python/04-agents/09_tree_of_thoughts.py
# Copyright 2026 Terrene Foundation
# SPDX-License-Identifier: Apache-2.0
"""
# ════════════════════════════════════════════════════════════════════════
# Kailash SDK Textbook — Agents / Tree-of-Thoughts Agent
# ════════════════════════════════════════════════════════════════════════
# OBJECTIVE: Use TreeOfThoughtsAgent for exploration-based reasoning
# LEVEL: Advanced
# PARITY: Python-only — no Rust equivalent
# VALIDATES: ToTAgent, ToTAgentConfig, ToTSignature, ToTPath,
#            ToTEvaluation, path generation, evaluation, selection
#
# Run: uv run python textbook/python/04-agents/09_tree_of_thoughts.py
# ════════════════════════════════════════════════════════════════════════
"""
from __future__ import annotations

import os

from dotenv import load_dotenv

load_dotenv()

from kaizen_agents.agents.specialized.tree_of_thoughts import (
    ToTAgent,
    ToTAgentConfig,
    ToTSignature,
    ToTPath,
    ToTEvaluation,
)
from kaizen.core.base_agent import BaseAgent

model = os.environ.get("DEFAULT_LLM_MODEL", "claude-sonnet-4-20250514")

# ── 1. Tree-of-Thoughts pattern ───────────────────────────────────────
# ToT explores multiple reasoning paths in parallel, then evaluates
# and selects the best one. This is fundamentally different from:
#
#   CoT:   Single linear reasoning path (one shot)
#   ReAct: Single iterative path with tool observations
#   ToT:   N parallel paths, evaluated, best selected
#
# The four phases:
#   1. GENERATE: Create N diverse reasoning paths
#   2. EVALUATE: Score each path independently
#   3. SELECT:   Choose the highest-scoring path
#   4. EXECUTE:  Extract the result from the best path

# ── 2. ToTPath and ToTEvaluation — typed structures ───────────────────
# TypedDict definitions provide type safety for paths and evaluations.

# ToTPath has required and optional fields
sample_path: ToTPath = {
    "path_id": 1,
    "reasoning": "Approach via cost-benefit analysis",
}
assert sample_path["path_id"] == 1
assert sample_path["reasoning"] == "Approach via cost-benefit analysis"

# ToTEvaluation scores a path
sample_eval: ToTEvaluation = {
    "path": {"path_id": 1, "reasoning": "Cost-benefit analysis"},
    "score": 0.85,
    "reasoning": "Well-structured approach with clear trade-offs",
}
assert sample_eval["score"] == 0.85

# ── 3. ToTSignature — multi-path I/O structure ────────────────────────

assert "task" in ToTSignature._signature_inputs

assert "paths" in ToTSignature._signature_outputs
assert "evaluations" in ToTSignature._signature_outputs
assert "best_path" in ToTSignature._signature_outputs
assert "final_result" in ToTSignature._signature_outputs

# ── 4. ToTAgentConfig — exploration-specific settings ─────────────────

config = ToTAgentConfig()

assert config.num_paths == 5, "Generate 5 reasoning paths by default"
assert config.max_paths == 20, "Safety limit: max 20 paths"
assert config.evaluation_criteria == "quality", "Default: quality-based evaluation"
assert config.parallel_execution is True, "Parallel generation by default"
assert config.temperature == 0.9, "High temperature for diverse paths"

# num_paths cannot exceed max_paths
try:
    ToTAgentConfig(num_paths=25, max_paths=20)
except ValueError:
    pass  # Config itself does not validate, but ToTAgent does

# Custom configuration
custom_config = ToTAgentConfig(
    num_paths=10,
    max_paths=20,
    evaluation_criteria="creativity",
    parallel_execution=True,
)
assert custom_config.num_paths == 10
assert custom_config.evaluation_criteria == "creativity"

# ── 5. ToTAgent instantiation ─────────────────────────────────────────
# ToT uses AsyncSingleShotStrategy (default) because the multi-path
# generation is handled internally, not via MultiCycleStrategy.
# Each path generation is a separate BaseAgent.run() call.

agent = ToTAgent(
    llm_provider="mock",
    model=model,
    num_paths=5,
    evaluation_criteria="quality",
)

assert isinstance(agent, ToTAgent)
assert isinstance(agent, BaseAgent)
assert agent.tot_config.num_paths == 5

# ── 6. Path validation (num_paths vs max_paths) ──────────────────────
# ToTAgent validates that num_paths does not exceed max_paths.

try:
    ToTAgent(
        llm_provider="mock",
        model=model,
        num_paths=25,
        max_paths=20,
    )
    assert False, "num_paths > max_paths should raise ValueError"
except ValueError as e:
    assert "exceeds" in str(e).lower()

# ── 7. Path evaluation logic ──────────────────────────────────────────
# The agent evaluates paths based on four criteria:
#   - Completeness: reasoning length > 20 chars (0.3)
#   - No errors: no "error" key in path (0.3)
#   - Structured steps: has non-empty steps list (0.2)
#   - Reasoning quality: reasoning > 100 chars (0.2)

# Good path: meets all criteria
good_path = {
    "path_id": 1,
    "reasoning": "A" * 150,  # Long reasoning
    "steps": ["step1", "step2", "step3"],
}
good_eval = agent._evaluate_path(good_path, "test task")
assert good_eval["score"] == 1.0, "Perfect path gets 1.0"

# Path with error: loses 0.3
error_path = {
    "path_id": 2,
    "reasoning": "A" * 150,
    "steps": ["step1"],
    "error": "generation failed",
}
error_eval = agent._evaluate_path(error_path, "test task")
assert error_eval["score"] == 0.7, "Error costs 0.3"

# Empty path: scores 0.0
empty_path = {"path_id": 3, "reasoning": ""}
empty_eval = agent._evaluate_path(empty_path, "test task")
assert empty_eval["score"] == 0.0, "Empty path scores 0.0"

# Short path: partial score
short_path = {"path_id": 4, "reasoning": "Brief reasoning about the problem"}
short_eval = agent._evaluate_path(short_path, "test task")
assert 0.0 < short_eval["score"] < 1.0, "Short path gets partial score"

# ── 8. Best path selection ────────────────────────────────────────────
# _select_best_path picks the evaluation with the highest score.

evaluations = [
    {"path": {"path_id": 1}, "score": 0.6, "reasoning": "OK"},
    {"path": {"path_id": 2}, "score": 0.9, "reasoning": "Excellent"},
    {"path": {"path_id": 3}, "score": 0.3, "reasoning": "Weak"},
]

best = agent._select_best_path(evaluations)
assert best["score"] == 0.9
assert best["path"]["path_id"] == 2

# Empty evaluations: graceful fallback
empty_best = agent._select_best_path([])
assert empty_best["score"] == 0.0

# ── 9. Input validation ───────────────────────────────────────────────

empty_result = agent.run(task="")
assert empty_result["error"] == "INVALID_INPUT"
assert empty_result["paths"] == []
assert empty_result["evaluations"] == []
assert empty_result["final_result"] == ""

# ── 10. Return structure ──────────────────────────────────────────────
# ToTAgent.run() returns:
# {
#     "paths": [
#         {"path_id": 0, "reasoning": "...", "steps": [...]},
#         {"path_id": 1, "reasoning": "...", "steps": [...]},
#         ...
#     ],
#     "evaluations": [
#         {"path": {...}, "score": 0.85, "reasoning": "..."},
#         ...
#     ],
#     "best_path": {"path": {...}, "score": 0.92, "reasoning": "..."},
#     "final_result": "The recommended approach is..."
# }
#
# Error codes: INVALID_INPUT, PATH_GENERATION_FAILED,
#              EVALUATION_FAILED, SELECTION_FAILED, EXECUTION_FAILED

# ── 11. Node metadata ─────────────────────────────────────────────────

assert ToTAgent.metadata.name == "ToTAgent"
assert "tot" in ToTAgent.metadata.tags
assert "multi-path" in ToTAgent.metadata.tags
assert "parallel" in ToTAgent.metadata.tags

print("PASS: 04-agents/09_tree_of_thoughts")

File: textbook/rust/04-agents/src/bin/09_tree_of_thoughts.rs
// Copyright 2026 Terrene Foundation
// SPDX-License-Identifier: Apache-2.0
//! Kailash SDK Textbook -- Agents / Tree of Thoughts
//!
//! OBJECTIVE: Build agents that explore multiple reasoning paths simultaneously.
//! LEVEL: Advanced
//! PARITY: Equivalent -- Python has TreeOfThoughts agent;
//!         Rust uses ToTConfig with parallel branch exploration.
//! VALIDATES: ToTConfig, BranchConfig, evaluation and pruning strategies,
//!            thought tree structure
//!
//! Run: cargo run -p tutorial-agents --bin 09_tree_of_thoughts

use kaizen_agents::agents::tree_of_thoughts::{
    BranchConfig, PruningStrategy, ToTConfig,
};

fn main() {
    // ── 1. Tree of Thoughts pattern ──
    // ToT extends Chain of Thought by exploring MULTIPLE reasoning paths:
    //   1. BRANCH: Generate N different approaches to the problem
    //   2. EVALUATE: Score each branch's quality and promise
    //   3. PRUNE: Discard low-scoring branches
    //   4. EXPAND: Continue the most promising branches
    //   5. SELECT: Choose the best final answer across all branches
    //
    // This is more expensive than CoT but excels at:
    //   - Creative problem-solving
    //   - Planning tasks
    //   - Complex puzzles
    //   - Tasks with multiple valid approaches

    // ── 2. ToTConfig ──
    // Configuration controls the tree exploration strategy.

    let config = ToTConfig::default();

    assert_eq!(config.num_branches(), 3); // Explore 3 paths
    assert_eq!(config.max_depth(), 3);    // Up to 3 levels deep
    assert!(matches!(
        config.pruning_strategy(),
        PruningStrategy::TopK
    ));

    // ── 3. Custom configuration ──

    let custom = ToTConfig::builder()
        .num_branches(5)
        .max_depth(4)
        .pruning_strategy(PruningStrategy::Threshold(0.6))
        .evaluation_prompt("Rate this approach from 0 to 1 for correctness and feasibility.")
        .build();

    assert_eq!(custom.num_branches(), 5);
    assert_eq!(custom.max_depth(), 4);

    // ── 4. BranchConfig ──
    // Each branch tracks its reasoning path and evaluation score.

    let branch = BranchConfig::new("approach_1")
        .thought("Start by breaking the problem into sub-problems")
        .score(0.8);

    assert_eq!(branch.name(), "approach_1");
    assert!((branch.score() - 0.8).abs() < 0.001);

    let branch2 = BranchConfig::new("approach_2")
        .thought("Apply a greedy algorithm to find a quick solution")
        .score(0.5);

    let branch3 = BranchConfig::new("approach_3")
        .thought("Use dynamic programming for optimal solution")
        .score(0.9);

    // ── 5. Pruning strategies ──

    // TopK: keep the top N branches by score
    assert!(matches!(
        ToTConfig::default().pruning_strategy(),
        PruningStrategy::TopK
    ));

    // Threshold: keep branches above a minimum score
    let threshold_config = ToTConfig::builder()
        .pruning_strategy(PruningStrategy::Threshold(0.6))
        .build();
    assert!(matches!(
        threshold_config.pruning_strategy(),
        PruningStrategy::Threshold(_)
    ));

    // ── 6. Tree structure ──
    // The tree is built level by level:
    //
    //   Level 0: Problem statement
    //   Level 1: [Branch A (0.8), Branch B (0.5), Branch C (0.9)]
    //            Branch B pruned (score < threshold)
    //   Level 2: [Branch A expanded (0.85), Branch C expanded (0.95)]
    //   Level 3: [Branch A final, Branch C final]
    //   Selection: Branch C chosen (highest cumulative score)

    // Simulate scoring and selection
    let branches = vec![branch, branch2, branch3];
    let best = branches
        .iter()
        .max_by(|a, b| a.score().partial_cmp(&b.score()).unwrap())
        .unwrap();
    assert_eq!(best.name(), "approach_3");

    // ── 7. Execution pattern ──
    //
    //   let agent = TreeOfThoughts::new(config, model).await;
    //   let result = agent.run("Plan a migration from monolith to microservices").await;
    //   println!("Best approach: {}", result["selected_branch"]);
    //   println!("Reasoning: {}", result["reasoning"]);
    //   println!("Branches explored: {}", result["branches_explored"]);
    //
    // NOTE: We do not call run() here as it requires an LLM API key.

    // ── 8. Key concepts ──
    // - ToT: explores multiple reasoning paths in parallel
    // - BranchConfig: individual path with thought and score
    // - PruningStrategy: TopK or Threshold for eliminating weak branches
    // - More expensive than CoT but better for complex/creative tasks
    // - Tree depth controls reasoning complexity
    // - LLM evaluates each branch (self-evaluation pattern)
    // - Final selection picks the highest-scoring complete path

    println!("PASS: 04-agents/09_tree_of_thoughts");
}
